central/role/sachelper/sachelper.go

```diff
@@ -0,0 +1,185 @@
+	query.Pagination = getSanitizedPagination(pagination, allowedSortFields)
+	results, err := h.clusterDataStore.Search(clusterLookupCtx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	optionsMap := getClustersOptionsMap()
+	clusters := extractScopeElements(results, optionsMap, search.Cluster.String())
+
+	return clusters, nil
+}
+
+func (h *sacHelperImpl) IsClusterVisibleForPermissions(
+	ctx context.Context,
+	clusterID string,
+	resourcesWithAccess []permissions.ResourceWithAccess,
+) (bool, error) {
+	clusterFound, hasFullAccess, err := hasClusterIDInScope(ctx, clusterID, resourcesWithAccess)
+	if err != nil {
+		return false, err
+	}
+	if hasFullAccess {
+		// Use an elevated context to check the existence of the cluster associated with the listed ID.
+		// This context must not be propagated.
+		elevatedCtx := sac.WithGlobalAccessScopeChecker(
+			ctx,
+			sac.AllowFixedScopes(
+				sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+				sac.ResourceScopeKeys(resources.Cluster),
+			),
+		)
+		return h.clusterDataStore.Exists(elevatedCtx, clusterID)
+	}
+	return clusterFound, nil
+}
+
+func (h *sacHelperImpl) GetNamespacesForClusterAndPermissions(
+	ctx context.Context,
+	clusterID string,
+	requestedPermissions []string,
+) ([]*v1.ScopeObject, error) {
+	resourcesWithAccess := listReadPermissions(requestedPermissions, permissions.NamespaceScope)
+	allNsResourcesWithAccess := listReadPermissions([]string{}, permissions.NamespaceScope)
+
+	clusterVisible, err := h.IsClusterVisibleForPermissions(ctx, clusterID, allNsResourcesWithAccess)
+	if err != nil {
+		return nil, err
+	}
+	if !clusterVisible {
+		return nil, errox.NotFound
+	}
+
+	namespacesInScope, hasFullAccess, err := listNamespaceNamesInScope(ctx, clusterID, resourcesWithAccess)
+	if err != nil {
+		return nil, err
+	}
+
+	// Use an elevated context to fetch namespace IDs and names associated with the listed namespace names.
+	// This context must not be propagated.
+	// The search is restricted to the namespace name field, and to the namespaces allowed
+	// by the extended scope.
+	var namespaceLookupCtx context.Context
+	if hasFullAccess {
+		namespaceLookupCtx = sac.WithGlobalAccessScopeChecker(ctx,
+			sac.AllowFixedScopes(
+				sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+				sac.ResourceScopeKeys(resources.Namespace),
+				sac.ClusterScopeKeys(clusterID),
+			),
+		)
+	} else {
+		namespaceLookupCtx = sac.WithGlobalAccessScopeChecker(ctx,
+			sac.AllowFixedScopes(
+				sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+				sac.ResourceScopeKeys(resources.Namespace),
+				sac.ClusterScopeKeys(clusterID),
+				sac.NamespaceScopeKeys(namespacesInScope.AsSlice()...),
+			),
+		)
+	}
+	query := search.NewQueryBuilder().
+		AddStringsHighlighted(search.Namespace, search.WildcardString).
+		ProtoQuery()
+	/*
+		// Currently, the namespace search overrides pagination information. As a consequence, pagination is disabled here.
```

Comment:
As this is leftover from the previous PR, did we have a ticket for this? Might be worthwhile referencing this here.